File: Info/Allumette/python/bot.py
from rules import *

WIN=1
LOST=-1
result = {}

def f(allu, myTurn=False):
	if allu <= 0:
		return(LOST if myTurn else WIN)
	if allu in result :
		if myTurn == result[allu][1]:
			return(result[allu][0])
		else :
			return(-result[allu][0])
	if myTurn:
		for value in RULES:
			result[allu - value] = f(allu-value,not myTurn), not myTurn
			if result[allu - value][0] == WIN:
				return(WIN)
		return(LOST)
	else :
		for value in RULES:
			result[allu - value] = f(allu-value,not myTurn), not myTurn
			if result[allu - value][0] == LOST:
				return(LOST)
		return(WIN)

def play(allu):
	for value in RULES:
		if f(allu-value) == WIN:
			return(value)
	return(1)

File: C/bot.c
#include "linked_list.h"
#include "display.h"
#include <stdlib.h>

#define WIN 1
#define LOST -1

linked_list *dico=NULL;

int	rec(int allu, int myTurn)
{
	int i = 1;
	int check = get_data(&dico,allu);
	int value;

	if (allu <= 0)
		return((myTurn) ? LOST : WIN);

	// if (check != 0)
	// {
	// 	if (myTurn)
	// 		return(check);
	// 	return(-check);
	// }

	if (myTurn)
	{
		while (i < 4)
		{
			value =rec(allu-i,!myTurn);
			set_data(&dico,allu-i,value);
			if (value == WIN)
				return(WIN);
			i++;
		}
		return(LOST);
	}
	else
	{
		while (i < 4)
		{
			value =rec(allu-i,!myTurn);
			set_data(&dico,allu-i,-value);
			if (value == LOST)
				return(LOST);
			i++;
		}
		return(WIN);
	}
}

int	bot(int allu)
{
	int i = 1;
	while (i < 4)
	{
		if (rec(allu - i, 0) == 1)
			return(i);
		i += 1;
	}
	return(1);
}
void disp(void)
{
	put_str("Test\n");
	put_nb(get_data(&dico,1));
	put_nb(get_data(&dico,2));
	put_nb(get_data(&dico,3));
	put_nb(get_data(&dico,4));
	put_nb(get_data(&dico,5));
	put_str("Test\n");
}
